packages/dds/tree/src/util/rangeMap.ts

```diff
@@ -6,229 +6,239 @@
+	public getAllEntries(): RangeQueryResult<T>[] {
+		return this.entries;
+	}
 
 	/**
-	 * The length of the prefix of the query range which have the same value.
-	 * For example, if a RangeMap has the same value for keys 5, 6, and 7,
-	 * a query about the range [5, 10] would give a result with length 3.
+	 * Retrieves the value for some prefix of the query range.
+	 *
+	 * @param start - The first key in the query range.
+	 * @param length - The length of the query range.
+	 * @returns A RangeQueryResult containing the value associated with `start`,
+	 * and the number of consecutive keys with that same value.
 	 */
-	length: number;
-}
+	public get(start: number, length: number): RangeQueryResult<T> {
+		for (const entry of this.entries) {
+			if (entry.start > start) {
+				return { value: undefined, length: Math.min(entry.start - start, length) };
+			}
 
-/**
- * See comments on `RangeQueryResult`.
- */
-export function getFromRangeMap<T>(
-	map: RangeMap<T>,
-	start: number,
-	length: number,
-): RangeQueryResult<T> {
-	for (const range of map) {
-		if (range.start > start) {
-			return { value: undefined, length: Math.min(range.start - start, length) };
+			const lastRangeKey = entry.start + entry.length - 1;
+			if (lastRangeKey >= start) {
+				const overlapLength = lastRangeKey - start + 1;
+				return { value: entry.value, length: Math.min(overlapLength, length) };
+			}
 		}
 
-		const lastRangeKey = range.start + range.length - 1;
-		if (lastRangeKey >= start) {
-			// This range contains `start`.
-			const overlapLength = lastRangeKey - start + 1;
-			return { value: range.value, length: Math.min(overlapLength, length) };
-		}
+		// There were no entries intersecting the query range, so the entire query range has undefined value.
+		return { value: undefined, length };
 	}
 
-	// There were no entries intersecting the query range, so the entire query range has undefined value.
-	return { value: undefined, length };
-}
+	/**
+	 * Sets the value for a specified range.
+	 *
+	 * @param start - The first key in the range being set.
+	 * @param length - The length of the range.
+	 * @param value - The value to associate with the range.
+	 */
+	public set(start: number, length: number, value: T): void {
+		const end = start + length - 1;
+		const newEntry: RangeEntry<T> = { start, length, value };
+
+		let iBefore = -1;
+		let iAfter = this.entries.length;
+		for (const [i, entry] of this.entries.entries()) {
+			const entryLastKey = entry.start + entry.length - 1;
+			if (entryLastKey < start) {
+				iBefore = i;
+			} else if (entry.start > end) {
+				iAfter = i;
+				break;
+			}
+		}
 
-export function getFirstEntryFromRangeMap<T>(
-	map: RangeMap<T>,
-	start: number,
-	length: number,
-): RangeEntry<T> | undefined {
-	const lastQueryKey = start + length - 1;
-	for (const range of map) {
-		if (range.start > lastQueryKey) {
-			// We've passed the end of the query range.
-			break;
+		if (value === undefined) {
```

Comment:
Consider moving this to the top of the function.

Reply:
Did you mean to change the signature of this method to explicitly accept `undefined` for the `value` parameter? Or are you just supporting the case where `undefined` is assignable to `T`?